samples/boards/nrf52/mesh/onoff-app/src/main.c

```diff
@@ -0,0 +1,772 @@
+#endif
+#if defined(CONFIG_BT_MESH_GATT_PROXY)
+	.gatt_proxy = BT_MESH_GATT_PROXY_ENABLED,
+#else
+	.gatt_proxy = BT_MESH_GATT_PROXY_NOT_SUPPORTED,
+#endif
+	.default_ttl = 7,
+
+	/* 3 transmissions with 20ms interval */
+	.net_transmit = BT_MESH_TRANSMIT(2, 20),
+	.relay_retransmit = BT_MESH_TRANSMIT(2, 20),
+};
+
+/*
+ * Client Configuration Declaration
+ */
+
+static struct bt_mesh_cfg_cli cfg_cli = {
+};
+
+/*
+ * Health Client/Server Declarations
+ */
+
+static struct bt_mesh_health_srv health_srv = {
+};
+
+static struct bt_mesh_health_srv health_cli = {
+};
+
+/*
+ * Publication Declarations
+ *
+ * The publication messages are initialized to the
+ * the size of the opcode + content + MIC
+ *
+ * The messages are in static storage because NET_BUF_SIMPLE()
+ * only allocates on the stack if called within a function.
+ * For publication, the message must be in static or global as
+ * it is re-transmitted several times. This occurs
+ * after the function that called bt_mesh_model_publish() has
+ * exited and the stack is no longer valid.
+ *
+ * Note that the additional 4 bytes for the AppMIC is not needed
+ * because it is added to a stack variable at the time a
+ * transmission occurs.
+ *
+ */
+
+static struct bt_mesh_model_pub health_pub = {
+	.msg  = BT_MESH_HEALTH_FAULT_MSG(0),
+};
+
+static struct bt_mesh_model_pub gen_level_pub;
+static struct bt_mesh_model_pub gen_onoff_pub_srv = {
+	.msg = NET_BUF_SIMPLE(2 + 2),
+};
+static struct bt_mesh_model_pub gen_onoff_pub_cli = {
+	.msg = NET_BUF_SIMPLE(2 + 2),
+};
+static struct bt_mesh_model_pub gen_onoff_pub_srv_s_0 = {
+	.msg = NET_BUF_SIMPLE(2 + 2),
+};
+static struct bt_mesh_model_pub gen_onoff_pub_cli_s_0 = {
+	.msg = NET_BUF_SIMPLE(2 + 2),
+};
+static struct bt_mesh_model_pub gen_onoff_pub_srv_s_1 = {
+	.msg = NET_BUF_SIMPLE(2 + 2),
+};
+static struct bt_mesh_model_pub gen_onoff_pub_cli_s_1 = {
+	.msg = NET_BUF_SIMPLE(2 + 2),
+};
+static struct bt_mesh_model_pub gen_onoff_pub_srv_s_2 = {
+	.msg = NET_BUF_SIMPLE(2 + 2),
+};
+static struct bt_mesh_model_pub gen_onoff_pub_cli_s_2 = {
+	.msg = NET_BUF_SIMPLE(2 + 2),
+};
+
+struct device *led_device[4], *sw_device;
+
+struct sw {
+	u32_t pin;
+	struct k_work button_work;
+};
+
+static struct sw sw;
+
+static struct gpio_callback button_cb;
+
+static int led_gpio_pin[] = {
+	LED0_GPIO_PIN,
+	LED1_GPIO_PIN,
+	LED2_GPIO_PIN,
+	LED3_GPIO_PIN,
+};
+
+static u8_t onoff_state[4] = {0};
+static u8_t last_onoff_state[4] = {1, 1, 1, 1};
+static u8_t level;
+static u8_t trans_id;
+static u32_t time, last_time;
+static u16_t primary_addr;
+static u16_t primary_net_idx;
+
+/*
+ * Generic OnOff Model Server Message Handlers
+ *
+ * Mesh Model Specification 3.1.1
+ *
+ */
+
+static void gen_onoff_get(struct bt_mesh_model *model,
+			  struct bt_mesh_msg_ctx *ctx,
+			  struct net_buf_simple *buf)
+{
+	struct net_buf_simple *msg = NET_BUF_SIMPLE(2 + 1 + 4);
+	u8_t elem_idx;
+
+	elem_idx = model->elem->addr - primary_addr;
```

Comment:
I wonder if it'd be possible to get rid of this slightly hackish way of looking up extra context for the element in question. The BT_MESH_MODEL() macro takes a last "user data" parameter which will then subsequently be available through model->user_data. You might need to reorganize how you store the various context & states to make this work out cleanly however.

Reply:
That's a great suggestion. 
I can tuck the current and last onoff states there.

Back to the drawing board.
Thanks

Reply:
So, I'd do something like this?
There's more state than this, but for example.
```
struct onoff_state {
        u8_t current;
        u8_t previous;
};

static struct onoff_state onoff_state[4];`

BT_MESH_MODEL(BT_MESH_MODEL_ID_GEN_ONOFF_SRV, gen_onoff_srv_op,
                      &gen_onoff_pub_srv, &onoff_state[0]),
BT_MESH_MODEL(BT_MESH_MODEL_ID_GEN_ONOFF_CLI, gen_onoff_cli_op,
                      &gen_onoff_pub_cli, &onoff_state[0]),
```
And to retrieve it in the function
```
struct onoff_state *onoff_state = (struct *onoff_state)model->user_data;

net_buf_simple_add_u8(msg, onoff_state->current);


```

Reply:
Yes, looks about right, however you don't need the explicit type cast when assigning to `onoff_state` from `model->user data`, since it's a void pointer.

Reply:
OK, I made the changes. I have the current and previous onoff states along with the led pin and device assignments attached to the model. It still works.

I reorganized it a bit and put the declarations at the beginning.